File: frontend/src/lib/productionCosting.ts
/**
 * Production Costing Engine — TypeScript types and API client.
 *
 * Costing data flows:
 *   MaterialConsumption × Material.standard_cost → total_material_cost
 *   LaborLog × WorkCenter.labor_rate_per_hour     → total_labor_cost
 *   WorkOrder duration × machine_cost_per_hour    → total_machine_cost
 *   UtilityTransaction (by batch_id)              → total_energy_cost
 */
import { apiClient } from "./api";

// ── Types ──────────────────────────────────────────────────────────────────────

export interface OrderCostBreakdown {
  order_id:                string;
  order_no:                string;
  product_name:            string | null;
  product_sku:             string | null;
  status:                  string;
  actual_quantity:         number | null;
  uom:                     string;

  total_material_cost:     number | null;
  total_labor_cost:        number | null;
  total_machine_cost:      number | null;
  total_energy_cost:       number | null;
  total_cost:              number | null;
  cost_per_unit:           number | null;
  standard_cost_per_unit:  number | null;
  cost_variance_pct:       number | null;
  costing_finalized_at:    string | null;

  material_row_count:      number | null;
  labor_row_count:         number | null;
}

export interface CostReportRow {
  product_id:              string;
  sku:                     string;
  product_name:            string;
  standard_cost_per_unit:  number;
  order_count:             number;
  total_qty:               number;
  total_material_cost:     number;
  total_labor_cost:        number;
  total_machine_cost:      number;
  total_energy_cost:       number;
  total_cost:              number;
  avg_cost_per_unit:       number;
  avg_variance_pct:        number;
  material_pct:            number;
  labor_pct:               number;
  machine_pct:             number;
  energy_pct:              number;
}

export interface CostTrendPoint {
  date:              string;
  material_cost:     number;
  labor_cost:        number;
  machine_cost:      number;
  energy_cost:       number;
  total_cost:        number;
  total_qty:         number;
  avg_cost_per_unit: number;
  order_count:       number;
}

export interface CostKPIs {
  order_count:           number;
  total_qty:             number;
  total_material_cost:   number;
  total_labor_cost:      number;
  total_machine_cost:    number;
  total_energy_cost:     number;
  total_cost:            number;
  avg_cost_per_unit:     number;
  avg_variance_pct:      number;
  over_budget_count:     number;
  material_pct:          number;
  labor_pct:             number;
  machine_pct:           number;
  energy_pct:            number;
  flag_high_variance:    boolean;
  flag_high_material:    boolean;
}

export interface CostFilters {
  date_from?:  string;
  date_to?:    string;
  product_id?: string;
}

export interface WIPRow {
  order_id:          string;
  order_no:          string;
  product_name:      string | null;
  product_sku:       string | null;
  planned_quantity:  number;
  actual_quantity:   number | null;
  uom:               string;
  status:            string;
  wip_material_cost: number;
  wip_labor_cost:    number;
  wip_machine_cost:  number;
  wip_total:         number;
  scheduled_start:   string | null;
  scheduled_end:     string | null;
}

export interface VarianceDetailRow {
  order_id:              string;
  order_no:              string;
  product_name:          string | null;
  product_sku:           string | null;
  status:                string;
  actual_quantity:       number | null;
  std_material_cost:     number;
  actual_material_cost:  number;
  material_variance:     number;
  std_labor_cost:        number;
  actual_labor_cost:     number;
  labor_variance:        number;
  std_machine_cost:      number;
  actual_machine_cost:   number;
  machine_variance:      number;
  total_variance:        number;
  variance_pct:          number | null;
}

export interface WorkCenterUtilRow {
  work_center_id:        string;
  work_center_name:      string;
  work_center_code:      string;
  capacity_hours:        number | null;
  actual_hours_used:     number;
  utilization_pct:       number | null;
  completed_orders:      number;
  labor_cost_incurred:   number;
  machine_cost_incurred: number;
}

// ── API client ─────────────────────────────────────────────────────────────────

export const productionCostApi = {
  async kpis(params?: Pick<CostFilters, "date_from" | "date_to">): Promise<CostKPIs> {
    const r = await apiClient.get<CostKPIs>("/api/v1/production-cost/kpis", { params });
    return r.data;
  },

  async report(params?: CostFilters): Promise<CostReportRow[]> {
    const r = await apiClient.get<CostReportRow[]>("/api/v1/production-cost/report", { params });
    return r.data;
  },

  async trend(params?: CostFilters): Promise<CostTrendPoint[]> {
    const r = await apiClient.get<CostTrendPoint[]>("/api/v1/production-cost/trend", { params });
    return r.data;
  },

  async getOrderCost(orderId: string): Promise<OrderCostBreakdown> {
    const r = await apiClient.get<OrderCostBreakdown>(
      `/api/v1/production-cost/orders/${orderId}/cost`,
    );
    return r.data;
  },

  async finalizeOrderCost(orderId: string): Promise<OrderCostBreakdown> {
    const r = await apiClient.post<OrderCostBreakdown>(
      `/api/v1/production-cost/orders/${orderId}/finalize`,
    );
    return r.data;
  },

  async wip(): Promise<WIPRow[]> {
    const r = await apiClient.get<WIPRow[]>("/api/v1/production-cost/wip");
    return r.data;
  },

  async varianceDetail(params?: Pick<CostFilters, "date_from" | "date_to">): Promise<VarianceDetailRow[]> {
    const r = await apiClient.get<VarianceDetailRow[]>("/api/v1/production-cost/variance-detail", { params });
    return r.data;
  },

  async workCenterUtilization(date_from: string, date_to: string): Promise<WorkCenterUtilRow[]> {
    const r = await apiClient.get<WorkCenterUtilRow[]>("/api/v1/production-cost/work-center-utilization", {
      params: { date_from, date_to },
    });
    return r.data;
  },
};

// ── Formatting helpers ──────────────────────────────────────────────────────────

export function fmtKES(v: number | null | undefined): string {
  if (v == null) return "—";
  return new Intl.NumberFormat("en-KE", {
    style: "currency", currency: "KES", maximumFractionDigits: 2,
  }).format(v);
}

export function fmtVariance(v: number | null | undefined): {
  text: string; color: string;
} {
  if (v == null) return { text: "—", color: "text-slate-400" };
  const sign = v > 0 ? "+" : "";
  const color = v > 10 ? "text-red-400" : v > 0 ? "text-amber-400" : "text-emerald-400";
  return { text: `${sign}${v.toFixed(1)}%`, color };
}

// Recharts pie data for cost breakdown
export function buildBreakdownPie(kpis: CostKPIs) {
  return [
    { name: "Materials",  value: kpis.total_material_cost, pct: kpis.material_pct, fill: "#6366f1" },
    { name: "Labor",      value: kpis.total_labor_cost,    pct: kpis.labor_pct,    fill: "#22d3ee" },
    { name: "Machine",    value: kpis.total_machine_cost,  pct: kpis.machine_pct,  fill: "#f59e0b" },
    { name: "Energy",     value: kpis.total_energy_cost,   pct: kpis.energy_pct,   fill: "#10b981" },
  ].filter(d => d.value > 0);
}

File: backend/app/schemas/production_costing.py
"""
Production Costing Schemas
"""
from __future__ import annotations
from typing import Optional, List
from pydantic import BaseModel, ConfigDict


class OrderCostBreakdown(BaseModel):
    """Detailed cost breakdown for a single production order."""
    model_config = ConfigDict(from_attributes=True)

    order_id:                str
    order_no:                str
    product_name:            Optional[str]   = None
    product_sku:             Optional[str]   = None
    status:                  str
    actual_quantity:         Optional[float] = None
    uom:                     str

    total_material_cost:     Optional[float] = None
    total_labor_cost:        Optional[float] = None
    total_machine_cost:      Optional[float] = None
    total_energy_cost:       Optional[float] = None
    total_cost:              Optional[float] = None
    cost_per_unit:           Optional[float] = None
    standard_cost_per_unit:  Optional[float] = None
    cost_variance_pct:       Optional[float] = None
    costing_finalized_at:    Optional[str]   = None

    # live-computed extras (not stored)
    material_row_count:      Optional[int]   = None
    labor_row_count:         Optional[int]   = None


class CostReportRow(BaseModel):
    """Aggregated cost row per product for the report."""
    product_id:              str
    sku:                     str
    product_name:            str
    standard_cost_per_unit:  float
    order_count:             int
    total_qty:               float
    total_material_cost:     float
    total_labor_cost:        float
    total_machine_cost:      float
    total_energy_cost:       float
    total_cost:              float
    avg_cost_per_unit:       float
    avg_variance_pct:        float
    material_pct:            float
    labor_pct:               float
    machine_pct:             float
    energy_pct:              float


class CostTrendPoint(BaseModel):
    """One day of cost trend data."""
    date:              str
    material_cost:     float
    labor_cost:        float
    machine_cost:      float
    energy_cost:       float
    total_cost:        float
    total_qty:         float
    avg_cost_per_unit: float
    order_count:       int


class CostKPIs(BaseModel):
    """Top-level cost KPIs for the dashboard."""
    order_count:           int
    total_qty:             float
    total_material_cost:   float
    total_labor_cost:      float
    total_machine_cost:    float
    total_energy_cost:     float
    total_cost:            float
    avg_cost_per_unit:     float
    avg_variance_pct:      float
    over_budget_count:     int
    material_pct:          float
    labor_pct:             float
    machine_pct:           float
    energy_pct:            float
    flag_high_variance:    bool
    flag_high_material:    bool


class WIPRow(BaseModel):
    """WIP value for one in-progress production order."""
    order_id:              str
    order_no:              str
    product_name:          Optional[str]   = None
    product_sku:           Optional[str]   = None
    planned_quantity:      float
    actual_quantity:       Optional[float] = None
    uom:                   str
    status:                str
    wip_material_cost:     float
    wip_labor_cost:        float
    wip_machine_cost:      float
    wip_total:             float
    scheduled_start:       Optional[str]   = None
    scheduled_end:         Optional[str]   = None


class VarianceDetailRow(BaseModel):
    """Component-level variance for one production order."""
    order_id:              str
    order_no:              str
    product_name:          Optional[str]   = None
    product_sku:           Optional[str]   = None
    status:                str
    actual_quantity:       Optional[float] = None
    # Material
    std_material_cost:     float
    actual_material_cost:  float
    material_variance:     float            # actual - standard (positive = over)
    # Labor
    std_labor_cost:        float
    actual_labor_cost:     float
    labor_variance:        float
    # Machine
    std_machine_cost:      float
    actual_machine_cost:   float
    machine_variance:      float
    # Total
    total_variance:        float
    variance_pct:          Optional[float]  = None


class WorkCenterUtilRow(BaseModel):
    """Utilization data for one work center over a date range."""
    work_center_id:        str
    work_center_name:      str
    work_center_code:      str
    capacity_hours:        Optional[float]  = None
    actual_hours_used:     float
    utilization_pct:       Optional[float]  = None
    completed_orders:      int
    labor_cost_incurred:   float
    machine_cost_incurred: float
